src/Build/Logging/BinaryLogger/BinaryLogger.cs

```diff
@@ -226,12 +227,21 @@ public void Shutdown()
 
             if (projectImportsCollector != null)
             {
+                projectImportsCollector.Close();
+
                 if (CollectProjectImports == ProjectImportsCollectionMode.Embed)
                 {
-                    eventArgsWriter.WriteBlob(BinaryLogRecordKind.ProjectImportArchive, projectImportsCollector.GetAllBytes());
+                    var archiveFilePath = projectImportsCollector.ArchiveFilePath;
+
+                    // It is possible that the archive couldn't be created for some reason.
+                    // Only embed it if it actually exists.
+                    if (FileSystems.Default.FileExists(archiveFilePath))
+                    {
+                        eventArgsWriter.WriteBlob(BinaryLogRecordKind.ProjectImportArchive, File.ReadAllBytes(archiveFilePath));
```

Comment:
I'm worried that this `ReadAllBytes` will still cause a big memory spike. Can we stream it from the disk into the log stream instead?

Reply:
Sure. I wanted to do this with more changes, but with proper planning - get rid of the temporary file completely by writing zip stream directly to binlog stream and using single reusable copy buffer for all files.
But removing `ReadAllBytes` required less changes than I expected and the result is pretty good :) I updated profiler result.

Reply:
> I updated profiler result.

Oh yeah, that looks beautiful now.